File: python/blowfish/blowfish.py
import argparse

def str2bin(str, length):
    if length == 64:
        assert len(str) == 8, 'length of the string is not 64 bits'
    elif length == 32:
        assert len(str) == 4, 'length of the string is not 64 bits'
    
    return ''.join('{:08b}'.format(ord(char)) for char in str)

def bin2hex(binary):
    return ''.join('{:x}'.format(int(binary[i:i + 4], 2)) for i in range(0, len(binary), 4))
    
def hex2bin(hex):
    return ''.join('{:04b}'.format(int(i, 16)) for i in hex)

def bin2int(binary):
    return int(binary, 2)

def bin2str(binary):
    return ''.join(chr(int(binary[i:i + 8], 2)) for i in range(0, len(binary), 8))

def hex2int(h):
    return int(h, 16)

def _xor(a: str, b: str):
    assert len(a) == 32 and len(b)== 32, 'lengths of the blocks does not match to 64 in XOR'
    return ''.join(str(int(i) ^ int(j)) for i, j in zip(a, b))

def _add(a: str, b: str):
    assert len(a) == 32 and len(b) == 32, 'lengths of blacks does not match to 32 in ADD'
    return '{:032b}'.format((bin2int(a) + bin2int(b)) % (2**32))

def _round(block, key, s_boxes):
    # This is an important part of the fiestel cipher where we establish
    # the left and right value, and which one well be xoring
    # before we flip the rolls and run through the fiestel network again
    assert len(block) == 64, 'Error in round input, len of block is not 64 bits'
    assert len(key) == 32, 'length of key is not 64 bits'
    L, R = block[:len(block) // 2], block[len(block) // 2]
    xored_L = _xor(L, key)
    R_prime = xored_L

    L_split = [bin2int(xored_L[i:i + 8]) for i in range(0, len(xored_L), 8)]
    assert len(L_split) == 4, 'The splitted L has no 4 elements'
    s_values = [s_boxes[i][L_split[i]] for i in range(4)]

    assert len(s_values[0]) == 32, 's_values are not of 32 bits'

    # add1(s3, xor1(s2, add(s1, s0)))
    result = _add(s_values[3], _xor(s_values[2], _add(s_values[1], s_values[0])))

    assert len(result) == 32, 'output of s_boxes operation is not 32 bits'
    L_prime = _xor(result, R)
    return L_prime + R_prime

def encrypt(block, p_array, s_blocks):
    for key in p_array[:16]:
        block = _round(block, key, s_blocks)
    
    left_block = block[:32]
    right_block = block[32:]
    right_block, left_block = left_block, right_block
    block = _xor(p_array[17], left_block) + _xor(p_array[16], right_block)
    return block

def generate_sub_keys(key):
    key = key * (72 // len(key)) + key
    key = key[:72]
    keys = [str2bin(key[i:i + 4], 32) for i in range(0, len(key), 4)]

    msg = '00000000'
    msg = str2bin(msg, 64)
    with open('pihex64k.txt', 'r') as f:
        P = [f.read(8) for i in range(18)]
        S = [[f.read(8) for j in range(256)] for k in range(4)]

    P = [hex2bin(i) for i in P]
    for i in range(4):
        for j in range(256):
            s[i][j] = hex2bin(S[i][j])

    P = [_xor(keys[i], P[i]) for i in range(18)]

    for i in range(0, len(P), 2):
        msg = encrypt(msg, P, S)
        P[i] = msg[:32]
        P[i + 1] = msg[32:]
    
    for i in range(4):
        for j in range(0, 256, 2):
            msg = encrypt(msg, P, S)
            S[i][j] = msg[:32]
            S[i][j + 1] = msg[32:]
    
    return P, S

def encryption(msg, sub_keys, s_boxes, mode):
    if mode == 'e':
        msg = [msg[i:i + 8] for i in range(0, len(msg), 8)]

        if len(msg[-1]) < 8:
            msg[-1] += ' ' * (8 - len(msg[-1]))
        
        msg = [str2bin(i, None) for i in msg]
    
    elif mode == 'd':
        msg = [hex2bin(msg[i:i + 16]) for i in range(0, len(msg), 16)]
    
    ciphertext = ''
    for snippet in msg:
        ciphertext += encrypt(snippet, sub_keys, s_boxes)
    
    if mode == 'e':
        cipher = bin2hex(ciphertext)

    if mode == 'd':
        cipher = bin2str(ciphertext)
    
    return cipher

parser = argparse.ArgumentParser()
parser.add_argument('-m', '--mode', choices=['e', 'd'], required=True, help='Encryption(e) / Decryption(d)')
parser.add_argument('-k', '--key', required=True, help='key for encryption or decryption')
parser.add_argument('-s', '--string', required=True, help='String to be encrypted or decrypted')
args = parser.parse_args()

message = args.string
password = args.key

sub_keys, s_boxes = generate_sub_keys(password)

if args.mode == 'e':
    enc = encryption(message, sub_keys, s_boxes, 'e')
    print('hex digest:', enc)
elif args.mode == 'd':
    dec = encryption(message, sub_keys[::-1], s_boxes, 'd')
    print(dec)
else:
    print('Invalid choice!')

File: javascript/blowfish/blowfish.js
const { ArgumentParser } = require('argparse');
const assert = require('assert');
const fs = require('fs');

const stringToBinary = (string, length) => {
    if (length === 64) {
        assert(string.length === 8, 'Length of the string is not 64 bits');
    } else if (length === 32) {
        assert(string.length === 4, 'Length of the string is not 64 bits');
    }

    const unicodeChars = [];

    for(let i = 0; i < string.length; i++) {
        unicodeChars.push(string.charCodeAt(i).toString(2).padStart(8, '0'));
    }

    return unicodeChars.join('');
}

const binaryToHex = (binary) => {
    const hexcodes = [];

    for(let i = 0; i < binary.length; i += 4) {
        hexcodes.push(parseInt(binary.slice(i, i + 4), 2).toString(16));
    }

    return hexcodes.join('');
}

const hexToBinary = (hex) => {
    const binaries = [];

    for(let i = 0; i < hex.length; i++) {
        binaries.push(parseInt(hex[i], 16).toString(2).padStart(4, '0'));
    }

    return binaries.join('');
}

const binaryToInt = (binary) => {
    return parseInt(binary, 2);
}

const binaryToString = (binary) => {
    let str = '';
    
    for(let i = 0; i < binary.length; i += 8) {
        const bin = binary.slice(i, i + 8);

        convertedBinary = String.fromCharCode(parseInt(bin, 2));
        str += convertedBinary;
    }

    return str;
}

const hexcodeToInteger = (hex) => {
    return parseInt(hex, 16);
}

const _xor = (a, b) => {
    assert(a.length === 32 && b.length === 32, 'lengths of the blocks do not match');

    let binary = '';
    for (let i = 0; i < a.length; i++) {
        const num = parseInt(a[i]) ^ parseInt(b[i]);
        binary = binary + num.toString();
    }
    return binary;
}

const _add = (a, b) => {
    assert(a.length === 32 && b.length === 32, 'Lengths of blocks do not match');

    const sum = binaryToInt(a) + binaryToInt(b);
    const mod = sum % (2**32);

    return mod.toString(2).padStart(32, '0');
}

const _round = (block, key, SBoxes) => {
    assert(block.length === 64, 'Error in round input, length of the block is not 64 bits');
    assert(key.length === 32, 'Length of key is not 32 bits');

    // Setup Left and Right blocks for the fiestel network
    const middle = block.length / 2;
    const L = block.slice(0, middle); 
    const R = block.slice(middle);

    // Flip
    const xoredLeft = _xor(L, key);
    const RightPrime = xoredLeft;

    const LeftSplit = [];

    for(let i = 0; i < xoredLeft.length; i += 8) {
        LeftSplit.push(binaryToInt(xoredLeft.slice(i, i + 8)));
    }

    assert(LeftSplit.length === 4, 'The splitted L has no 4 elements');

    const SValues = [];

    for(let i = 0; i < 4; i++) {
        SValues.push(SBoxes[i][LeftSplit[i]])
    };

    assert(SValues[0].length === 32, 'SValues are not 32 bits');

    const result = _add(SValues[3], _xor(SValues[2], _add(SValues[1], SValues[0])));

    assert(result.length === 32, 'Output of SBoxes operation is not 32 bits');

    const LeftPrime = _xor(result, R);

    return LeftPrime + RightPrime;
}

const encrypt = (block, PArray, SBlocks) => {
    for (let i = 0; i < 16; i++) {
        blockTemp = _round(block, PArray[i], SBlocks);
        block = blockTemp;
    };

    let leftBlock = block.slice(0, 32);
    let rightBlock = block.slice(32);
    // Switch the values
    let temp = leftBlock;
    leftBlock = rightBlock;
    rightBlock = temp;

    return _xor(PArray[17], leftBlock) + _xor(PArray[16], rightBlock);
}

const generateSubKeys = (key) => {
    // Repeating our key until we have a 72 byte long key
    // Which means our password length can only be 72 characters
    key = key.repeat(72 / key.length) + key;
    key = key.slice(0, 72);
    const keys = [];

    for(let i = 0; i < key.length; i += 4) {
        keys.push(stringToBinary(key.slice(i, i + 4), 32))
    }

    let P = [];
    let S = [];
    let msg = '00000000';
    msg = stringToBinary(msg, 64);

    // Read the txt file
    const data = fs.readFileSync('./pihex64k.txt', 'utf8');
    
    // i, j -> counters
    let i = 0; 
    let j = 0;
    while(i < 18) {
        P.push(data.slice(j, j + 8));
        i+= 1;
        j+= 8;
    };

    // x, y -> counters
    let x = 0;
    // Reuse the j -> counter to read file from where we left off
    while(x < 4) {
        let arr = [];
        let y = 0;
        while(y < 256) {  
            arr.push(data.slice(j, j + 8));
            y += 1;
            j += 8;
        }

        S.push(arr);
        x += 1;
    };

    // Convert all the hexes in P to binary
    P = P.map((hex) => hexToBinary(hex))

    for(let i = 0; i < 4; i++) {
        for(let j = 0; j < 256; j++) {
            S[i][j] = hexToBinary(S[i][j])
        };
    };
    
    // xor the 32 bit blocks in P with the keys
    P = P.map((ele, i) => {
        return _xor(keys[i], ele)
    });

    for(let i = 0; i < P.length; i += 2) {
        msg = encrypt(msg, P, S);
        P[i] = msg.slice(0, 32);
        P[i + 1] = msg.slice(32);
    }

    for(let i = 0; i < 4; i++) {
        for(let j = 0; j < 256; j += 2) {
            msg = encrypt(msg, P, S);
            S[i][j] = msg.slice(0, 32);
            S[i][j + 1] = msg.slice(32);
        }
    }

    return [P, S];
}

const encryption = (msg, subKeys, SBoxes, mode) => {
    const values = [];

    if(mode === 'e') {

        for(let i = 0; i < msg.length; i += 8) {
            values.push(msg.slice(i, i + 8))
        }

        const len = values.length - 1;

        if(values[len].length < 8) {
            values[len] = values[len].padEnd(8, ' ');
        }
        msg = values.map(ele => stringToBinary(ele));

    } else if (mode === 'd') {

        for(let i = 0; i < msg.length; i += 16) {
            values.push(hexToBinary(msg.slice(i, i + 16)));
        }
        msg = values;
    }

    let cipherText = '';
    let cipher;
    
    for (snippet of msg) {
        cipherText += encrypt(snippet, subKeys, SBoxes);
    }

    if (mode === 'e') {
        cipher = binaryToHex(cipherText);
    }

    if (mode === 'd') {
        cipher = binaryToString(cipherText);
    }

    return cipher
}

const parser = new ArgumentParser({})

parser.add_argument('-m', '--mode', { choices: ['e', 'd'], required: true, help: 'Encryption(e) / Decryption(d)'});
parser.add_argument('-k', '--key', { required: true, help: 'key for encryption or decryption' });
parser.add_argument('-s', '--string', { required: true, help: 'String to be encrypted or decrypted' });

const args = parser.parse_args();

const message = args.string;
const password = args.key;

const subKeys = generateSubKeys(password);

if(args.mode === 'e') {
    let enc = encryption(message, subKeys[0], subKeys[1], 'e');
    console.log('hex digest:', enc);
} else if(args.mode === 'd') {
    let dec = encryption(message, subKeys[0].reverse(), subKeys[1], 'd');
    console.log(dec);
} else {
    console.log('Invalid Choice');
}

/**
 * EASY TEST CASES BELOW
 */

// let binary = stringToBinary('aassddff', 64);
// let hexcode = binaryToHex(binary);
// let backToBin = hexToBinary('d1310ba6');
// let binaryNumber = binaryToInt(binary);
// let string = binaryToString(binary);
// let int = hexcodeToInteger(hexcode);
// // console.log()
// let a = '10010011111001010011100011101000';
// let b = '11100000101101110000110111001110';
// let x = _xor(a, b)
// let add = _add(a, b)
// let block = '1010000011011110011111010111000000110110101001000000110010110101';
// let key = '00001010110000100011010100001101';
// let gen = generateSubKeys('password')
// console.log(gen)
// console.log(backToBin, 'd1310ba6')




